### PR TITLE
Implement Objective-C nullability

### PR DESCRIPTION
**Pull request by Lucashuang0802**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/.github/CONTRIBUTING.md). Confirmation: ____

## What's in this pull request?

This PR adds the nullability modifiers in Objective-C properties, headers, and method declaration so `Optional Type` is available in Swift version. 

One thing to notice is all the warnings from the test target. Because we enable this type checking ability, some edge-case testings like checking `nil` warn that it's not supposed to be `nil`.

### CODE REVIEW

**Comment by Lucashuang0802** [comment]
The second commit is to update my fork from the main repo.

**Comment by Lucashuang0802** [comment]
Failed on validation of CocoaPods.

**Comment by jessesquires** [comment]
Thanks @Lucashuang0802 ! 👍 Please rebase on develop 😄

File: JSQMessagesViewController/Categories/NSString+JSQMessages.h [hunk 1]
```diff
@@ -23,6 +23,6 @@
 /**
  *  @return A copy of the receiver with all leading and trailing whitespace removed.
  */
-- (NSString *)jsq_stringByTrimingWhitespace;
+- (nonnull NSString *)jsq_stringByTrimingWhitespace;
 
 @end
```
Comment (jessesquires) [inline JSQMessagesViewController/Categories/NSString+JSQMessages.h]: let's wrap this with `NS_ASSUME_NONNULL_*`

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.h [hunk 2]
```diff
@@ -162,7 +163,7 @@
  *  you should also override `messagesViewController` to return your
  *  view controller loaded from your custom nib.
  */
-+ (UINib *)nib;
++ (nullable UINib *)nib;
```
Comment (jessesquires) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.h]: so.. some of this is non-intuitive.

we should actually remove `nullable` here. it should be nonnull, and we should update the docs to remove mention of `nil`.

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.h [hunk 3]
```diff
@@ -171,7 +172,7 @@
  *
  *  @return An initialized `JSQMessagesViewController` object if successful, `nil` otherwise.
  */
-+ (instancetype)messagesViewController;
++ (nullable instancetype)messagesViewController;
```
Comment (jessesquires) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.h]: same here, nonnull

File: JSQMessagesViewController/Controllers/JSQMessagesViewController.h [hunk 4]
```diff
@@ -303,3 +304,4 @@
 - (void)didReceiveMenuWillHideNotification:(NSNotification *)notification;
 
 @end
+NS_ASSUME_NONNULL_END
```
Comment (jessesquires) [inline JSQMessagesViewController/Controllers/JSQMessagesViewController.h]: always add newlines at end of files

File: JSQMessagesViewController/Categories/UIView+JSQMessages.h [hunk 5]
```diff
@@ -36,3 +37,4 @@
 - (void)jsq_pinAllEdgesOfSubview:(UIView *)subview;
 
 @end
+NS_ASSUME_NONNULL_END
```
Comment (jessesquires) [inline JSQMessagesViewController/Categories/UIView+JSQMessages.h]: always add newlines at end of files

File: JSQMessagesViewController/Categories/UIColor+JSQMessages.h [hunk 6]
```diff
@@ -53,4 +54,5 @@
  */
 - (UIColor *)jsq_colorByDarkeningColorWithValue:(CGFloat)value;
 
-@end
\ No newline at end of file
+@end
+NS_ASSUME_NONNULL_END
```
Comment (jessesquires) [inline JSQMessagesViewController/Categories/UIColor+JSQMessages.h]: newline

File: JSQMessagesViewController/Categories/UIImage+JSQMessages.h [hunk 7]
```diff
@@ -82,3 +83,4 @@
 + (UIImage *)jsq_defaultPauseImage;
 
 @end
+NS_ASSUME_NONNULL_END
```
Comment (jessesquires) [inline JSQMessagesViewController/Categories/UIImage+JSQMessages.h]: newline

File: JSQMessagesViewController/Categories/NSBundle+JSQMessages.h [hunk 8]
```diff
@@ -37,6 +38,7 @@
  *
  *  @return A localized version of the string designated by key in the JSQMessages table.
  */
-+ (NSString *)jsq_localizedStringForKey:(NSString *)key;
++ (nullable NSString *)jsq_localizedStringForKey:(NSString *)key;
 
 @end
+NS_ASSUME_NONNULL_END
```
Comment (jessesquires) [inline JSQMessagesViewController/Categories/NSBundle+JSQMessages.h]: newline

File: JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h [hunk 9]
```diff
@@ -36,7 +38,7 @@
 *
 *  @return An initialized `JSQMessagesAvatarImage` object if created successfully, `nil` otherwise.
 */
-+ (JSQMessagesAvatarImage *)avatarImageWithPlaceholder:(UIImage *)placeholderImage diameter:(NSUInteger)diameter;
++ (nullable JSQMessagesAvatarImage *)avatarImageWithPlaceholder:(UIImage *)placeholderImage diameter:(NSUInteger)diameter;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h]: same as above, should actually be nonnull. let's remove nullable here and update docs.

File: JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h [hunk 10]
```diff
@@ -49,7 +51,7 @@
  *
  *  @return An initialized `JSQMessagesAvatarImage` object if created successfully, `nil` otherwise.
  */
-+ (JSQMessagesAvatarImage *)avatarImageWithImage:(UIImage *)image diameter:(NSUInteger)diameter;
++ (nullable JSQMessagesAvatarImage *)avatarImageWithImage:(UIImage *)image diameter:(NSUInteger)diameter;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h]: same, nonnull

File: JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h [hunk 11]
```diff
@@ -59,7 +61,7 @@
  *
  *  @return A new image object if successful, `nil` otherwise.
  */
-+ (UIImage *)circularAvatarImage:(UIImage *)image withDiameter:(NSUInteger)diameter;
++ (nullable UIImage *)circularAvatarImage:(UIImage *)image withDiameter:(NSUInteger)diameter;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h]: same, nonnull

File: JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h [hunk 12]
```diff
@@ -70,7 +72,7 @@
  *
  *  @return A new image object if successful, `nil` otherwise.
  */
-+ (UIImage *)circularAvatarHighlightedImage:(UIImage *)image withDiameter:(NSUInteger)diameter;
++ (nullable UIImage *)circularAvatarHighlightedImage:(UIImage *)image withDiameter:(NSUInteger)diameter;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h]: same, nonnull

File: JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h [hunk 13]
```diff
@@ -90,10 +92,11 @@
  *  However, a font size `30.0f` and diameter of `10.0f` will not produce a desirable image.
  *  Further, this method does not check the length of userInitials. It is recommended that you pass a string of length `2` or `3`.
  */
-+ (JSQMessagesAvatarImage *)avatarImageWithUserInitials:(NSString *)userInitials
-                                        backgroundColor:(UIColor *)backgroundColor
-                                              textColor:(UIColor *)textColor
-                                                   font:(UIFont *)font
-                                               diameter:(NSUInteger)diameter;
++ (nullable JSQMessagesAvatarImage *)avatarImageWithUserInitials:(NSString *)userInitials
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesAvatarImageFactory.h]: same, nonnull

**Comment by jessesquires** [comment]
Hey @Lucashuang0802 😄 

Left a few comments and started to see a pattern. Let's:

- make sure we have newlines at the end of each file
- initializers should **not** be `nullable` unless you explicitly see them return `nil`. We should update the docs to remove mentions of `nil`. Sorry for the confusion here. 😅 

The original attempt at #1213 is a good reference.

**Comment by Lucashuang0802** [comment]
Yes, that was my fault that did not pay too many attentions on that. I will adjust it. Except those places you pointed out, I will use one of the most scary way(eye-scan) to check the header files again. lol.

Regarding to the docs change, do you prefer to include it in this PR or create a separate issue as parts of the 8.0 release?

**Comment by jessesquires** [comment]
No worries!

Yeah - let's include everything in this PR :smile:

**Comment by Lucashuang0802** [comment]
I did not squash just so it's easier for squash. I can squash it if it's okay to merge.

**Comment by jessesquires** [comment]
Awesome! Thanks @Lucashuang0802 ! 👍 

That last commit looks good. A couple more things:

1. anything `weak` **should** also be `nullable` (I saw a few of these were incorrect) The static analyzer in Xcode isn't smart enough to warn/error about this, but it is a bug. (`weak` reference can be nil'd at anytime)
2. Do you mind updating `CHANGELOG.md`? We can add this under `8.0` "Enhancements". Just follow the existing pattern 😄

**Comment by jessesquires** [comment]
Oh, and don't worry about squashing, we'll let GitHub do that 😊

**Comment by Lucashuang0802** [comment]
Definitely a silly change from a tired brain! It's updated it. Let me know if you find anything else. 😄

**Comment by Lucashuang0802** [comment]
The fail on CocoaPods validation:
```
The following build commands failed:
	CompileC /Users/lucas/Library/Developer/Xcode/DerivedData/App-gdnuvudsobxxxqdbmrziyrljxkey/Build/Intermediates/Pods.build/Release-iphonesimulator/JSQMessagesViewController.build/Objects-normal/i386/JSQSystemSoundPlayer+JSQMessages.o JSQMessagesViewController/JSQMessagesViewController/Categories/JSQSystemSoundPlayer+JSQMessages.m normal i386 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(1 failure)
 -> JSQMessagesViewController (7.3.2)
    - WARN  | url: There was a problem validating the URL http://jessesquires.github.io/JSQMessagesViewController.
    - WARN  | url: There was a problem validating the URL https://raw.githubusercontent.com/jessesquires/JSQMessagesViewController/develop/Screenshots/screenshot0.png.
    - WARN  | url: There was a problem validating the URL https://raw.githubusercontent.com/jessesquires/JSQMessagesViewController/develop/Screenshots/screenshot1.png.
    - WARN  | url: There was a problem validating the URL https://raw.githubusercontent.com/jessesquires/JSQMessagesViewController/develop/Screenshots/screenshot2.png.
    - WARN  | url: There was a problem validating the URL https://raw.githubusercontent.com/jessesquires/JSQMessagesViewController/develop/Screenshots/screenshot3.png.
    - WARN  | url: There was a problem validating the URL https://twitter.com/jesse_squires.
    - ERROR | [iOS] xcodebuild: Returned an unsuccessful exit code.
    - NOTE  | [iOS] xcodebuild:  JSQMessagesViewController/JSQMessagesViewController/Categories/JSQSystemSoundPlayer+JSQMessages.h:19:9: fatal error: 'JSQSystemSoundPlayer/JSQSystemSoundPlayer.h' file not found

Analyzed 1 podspec.

[!] The spec did not pass validation, due to 1 error.
```

`pod lib lint JSQMessagesViewController.podspec --verbose --allow-warnings` passed

**Comment by jessesquires** [comment]
Thanks @Lucashuang0802 ! 🎉

**Comment by MacMeDan** [comment]
Ahh it isn't merged yet. I will watch out for this when it does. 👍🏼

**Comment by Lucashuang0802** [comment]
Yeah, I noticed that you guys need help from pushing the Swift demo? Any places I should look into?

**Comment by MacMeDan** [comment]
We are just working to get it close to parity with the objective c demo. There are a few pull request waiting to be merged but anything else that you see not there from the objective C version could be added. Currently different types of media are being added to the example. What we have so far : is a group conversation and regular chat pull requested. I believe location media messages was just merged. There is tailless and tailed images added to settings and showing and hiding avatars currently waiting. 

What still need:
photo messages, video and the initial screen could use some help. It needs to be similar to the other example demonstrating different ways of instantiating.

**Comment by Lucashuang0802** [comment]
@MacMeDan I have some time this week. Do you want it to be like the demo in Objective-C?

**Comment by MacMeDan** [comment]
Ya that is what we are trying to move towards. we are keeping track of it in #1608
